### DISCUSSION TITLE
Next assets serving not working even with an assetPrefix

### DISCUSSION THREAD

**Discussion by Sletheren** [vercel/next.js#12131] [Help]
Hello there,
I have a NextJS application deployed on AWS and orchestrated using Kubernetes 
The logic of the app is it can be accessed from several URLs like this:
```
https://www.domain.com/A1/myApp1
https://www.domain.com/myApp1
https://www.domain.com/A2/myApp2
https://www.domain.com/myApp2
https://www.domain.com/foo/bar/myApp
...etc
```

The problem is NextJS tries to always load the assets from the root `https://www.domain.com/`\
even by using a dynamic assetPrefix like follows in my `server.js`

```
  server.get('*', (req, res) => {
    const reqUrl = req.protocol + '://' + req.headers.host + req.url
    const assetPrefix = loadAssetPrefix(reqUrl)
    app.setAssetPrefix(assetPrefix)
    return handle(req, res)
  })
```

and this results in having the app not being accessible and getting a load of 404's
like in the screenshot below:

https://prnt.sc/s4iu9r

I suspect it's something with the NEXTJS routing or serving assets outside the root of the server where the app is deployed

## Explanation

**Scenario 1:** 
When accessing the app using: **`https://www.domain.com/A1/myApp1`** without **assetPrefix**
the assets result in a 404 when the browser is trying to access: `https://www.domain.com/_next/static/development/pages/_app.js`

**Scenario 2:** 
When accessing the app using: **`https://www.domain.com/A1/myApp1`** with an **assetPrefix='/A1/myApp1'**
the assets result in a 404 as well when the browser is trying to access: `https://www.domain.com/A1/myApp1/_next/static/development/pages/_app.js`

## To Reproduce

Just simply have a nextJS app using the above server.js and try to host it in a subdirectory NOT in the root of your server,
You will end up with all your assets not being found whether using assetPrefix or not.

## Screenshots

My server.js file:
https://prnt.sc/s4is8j

My next.config.js:
https://prnt.sc/s4iss9

## System information

- OS: MacOS
- Version of Next.js: [e.g. 9.2.1]
- Version of Node.js: [e.g. 12.16.0]

**Comment by timneutkens** [comment]
I'm assuming you have some static pages and setting a dynamic assetPrefix doesn't work for those. Overall the assetPrefix has to be known at build time so that Next.js can generate the correct html output at build time for static pages.

**Reply by Sletheren** [reply]
Then how can I handle this use case when having multiple `assetPrefixes` ? as I mentioned in the example the app can be accessible from different `Urls`
What do you suggest? since I was really stuck with this it has been a week so far :/

**Reply by timneutkens** [reply]
Also note that assetPrefix only changes the location that is rendered, not where the assets are served from, so in the scenario that you're outlining you have to remove the additional prefix from `req.url` before routing to Next.js.

Also based on your example it doesn't seem like you particularly need the assets to live in multiple subroots, eg you can have a constant assetPrefix for all apps as it's the same application right?

**Reply by Sletheren** [reply]
Yes that is correct, It's just one app in the end,
Then that means in my `catch-all` middleware I just mutate the req.url and strip down `/A1/myApp` or `/myApp`per example, then pass the req to the nextjs handle method?

**Reply by timneutkens** [reply]
Why do you need to set the assetPrefix in that case to begin with? If you use the default behavior all routes will request `domain.com/_next/...etc` That's also better overall as users will use the same JS file across routes.

**Reply by Sletheren** [reply]
at first I thought `assetPrefix` it to let next know where to look for the `/_next/,,,etc` because it always look for it from the root as you mentioned
because if I guess now `domain.com/A1/myApp` the assets are loaded from `domain.com/_next` which basically doesn't exist because the app is not available from that domain

**Reply by timneutkens** [reply]
You can also set it to a consistent place and strip the subroute from `req.url`.

**Reply by Sletheren** [reply]
@timneutkens So what I did, is I created a middleware that strip the subroute for the request like follows:
```
server.use(function (req, res, next) {
    const reqUrl = req.protocol + '://' + req.headers.host + req.url
    const brandConfig = extractBrand(reqUrl)
    const subroute = brandConfig.subroute
    app.setAssetPrefix(subroute)
    req.originalUrl = req.originalUrl.replace(subroute, '') || '/'
    req.path = req.path.replace(subroute, '') || '/'
    req.url = req.url.replace(subroute, '') || '/'
    return next()
  })
```
When accessing: `www.domain.com/A1/MyApp/` then `subroute = /A1/MyApp` and after rewriting the URL I still get 404s for static assets:

`https://www.domain.com/static/images/logo.svg`

So the URL has been successfully rewritten but still Next is unable to find the assets inside /public/static

**Comment by Sletheren** [comment]
@timneutkens hello again,
so I tried the above work-arounds and fixes, 
It could only work with a combination of a middleware to rewrite the request URL and assetPrefix to make the assets accessible, but also I needed to add to all my assets something like this \
`<img src='${assetPrefix}/static/images/logo.svg' />` where assetPrefix equals the subdirectory where the app is hosted `/A1/MyApp` in my case

My question is, is there a way to go around this? or a better solution?
Thank you

**Comment by Jaynam07** [comment]
I am also facing similar issue, what can be done here ?

**Comment by fcano-ut** [comment]
I bumped into the same issue. We don't use a custom server, but we use Emissary Ingress inside our Kubernetes Cluster to map certain routes to our NextJS app. Emissary Ingress is configured with a path rewrite which allows you to have a route like `/my-app/a/b` being served by the app container, but mapped internally to `/a/b`. This works mostly great, but at the moment we start using `getServerSideProps` in any route, it breaks.

So in conclusion the new info I can provide about this bug:
- This error happens even if you're not using a custom Next.js server
- It starts happening when you use `getServerSideProps` in a route -- static routes work fine
